keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,309 @@
+
+#### Story 2
+
+A cluster orchestrator can design and implement a centralized pod eviction manager that handles all pod eviction and deletions on different `NoExecute`  taints (`Unreachable`, `NotReady` and others) and any custom taints, e.g., acting on the conditions defined in `descheduler`, such as node overloading, etc.
+
+**Story 3**
+
+A custom taint manager can make calls to other controllers or external systems to make a better decision whether, when and how to delete running pods on the tainted nodes (e.g. to prevent quorum loss and/or support different mechanism than Pod Disruption Budget to calculate applications availability).
+
+#### Story 4
+
+A pod may be stuck in terminating state for various reason, e.g.,  `kubelet` is not up and `AppServer` cannot reach `kubelet`, etc. The hanging deletion may prevent new pods starting because the persistent volume claim is not released. A custom taint-manager can perform a forceful deletion as needed.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+## Design Details
+
+### Current implementation
+
+In this current version, `NoExecuteTaintManager` is part of `NodeLifycleController`. It’s created and run if `enable-taint-manager` flag is true.  The flag is deprecated and will be removed in kubernetes 1.27. 
+
+<p align="center"><image src="current-nodelifecycle-controller.png" width=50%></p>
+
+### Proposed Controllers
+
+The proposed design refactors `NodeLifecycleController` into two independent controllers, which are managed by `kube-controller-manager`.
+
+1. `NodeLifecycleController` monitors node health and adds `NotReady` and `Unreachable` taints to nodes 
+2. `NoExecuteTaintManager` watches for node and pod updates,  and performs `NoExecute` taint based pod eviction
+
+The existing  kube-controller-manager flag `--controllers` will be used to optionally disable the `NoExecuteTaintManager`.
+
+<p align="center"><image src="new-controllers.png" width=60%></p>
+
+### Custom taint-manager
+
+Opting out from the default `TaintManger`, Kubernetes operators will be able to implement custom logic to handle `NoExecute` taint such as delegating the authority for the termination for some classes of workloads providing high-availability protection guarantees. They can also scale out the usage of `NoExcute` to more use-cases than `NotReady` nodes (e.g. unhealthy network, etc.).
+
+<p align="center"><image src="custom-taint-manager.png" width=90%></p>
+
+### Implementation
+
+A new `NodeLifeCycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.
+
+A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.  A new flag is added 
```

Comment:
We shouldn't say "a new flag", do we? as it will be a top-level controller controlled by `--controllers`.

```suggestion
A new `NoExecuteTaintManager` is created as a first-class controller managed by `kube-controller-manager`. Its implementation is based on the current taint-manager `controller/nodelifecycle/taint-manager.go`.
```